x/meshsecurity/keeper/stake.go

```diff
@@ -2,10 +2,66 @@ package keeper
 
 import (
 	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/cosmos/cosmos-sdk/types/errors"
+	stakingtypes "github.com/cosmos/cosmos-sdk/x/staking/types"
+
+	types "github.com/osmosis-labs/mesh-security-sdk/x/meshsecurity/types"
 )
 
-func (k Keeper) Delegate(pCtx sdk.Context, actor sdk.AccAddress, valAddr sdk.ValAddress, amt sdk.Coin) error {
-	panic("not implemented, yet")
+// Delegate delegate mints new "virtual" bonding tokens and delegates them to the given validator.
```

Comment:
Function looks very solid and nice idea with the cacheWrap to make all changes atomic (although the whole call would probably be in such a CacheWrap anyway, right?

Reply:
When triggered by external TX then this is always in a cache wrap. Not necessarily when called by a begin/endblock.
The cache wrap should always be at a higher level to cover the full process and not the method only but I copied this from Osmosis as an extra safety net.